pkg/expression/builtin_time.go

```diff
@@ -67,7 +67,34 @@ var (
 	durationPattern = regexp.MustCompile(`^\s*[-]?(((\d{1,2}\s+)?0*\d{0,3}(:0*\d{1,2}){0,2})|(\d{1,7}))?(\.\d*)?\s*$`)
 
 	// timestampPattern checks whether a string matches the format of timestamp.
-	timestampPattern = regexp.MustCompile(`^\s*0*\d{1,4}([^\d]0*\d{1,2}){2}\s+(0*\d{0,2}([^\d]0*\d{1,2}){2})?(\.\d*)?\s*$`)
+	timestampPattern = regexp.MustCompile(`^` +
+		// Skip any spaces or zeros
+		`\s*0*` +
+		// Year 1-4 digits
+		`\d{1,4}` +
+		// TODO: Add warning if non '-' separator in ParseTime
+		// 1 or 2 digit Month and Day
+		// Any non-digit as separator
+		// Any leading 0's for Month/Day
+		`([^\d]0*\d{1,2}){2}` +
+		// At least one space between Date and Time parts
+		`\s+` +
+		// Hour is mandatory
+		// Any number of leading zeroes
+		// 1-2 Hour digits
+		`0*\d{1,2}` +
+		// Minutes or Minutes:Seconds are optional
+		// Any non-digit separator before Minute and Second parts
+		// Any number of leading zeroes in Min/Sec!
+		// 1-2 digit minutes/seconds
+		`([^\d]0*\d{1,2}){0,2}` +
```

Comment:
Another fix for allowing `SELECT TIMESTAMP '2024-01-01 18[:00]'`. `SELECT TIMESTAMP '2024-01-01 18:00:00'` did already work.